[FLINK-23843][runtime] Properly fail the job when SplitEnumeratorContext.runInCoordinatorThread() throws an exception

echauchot opened the pull request:
## What is the purpose of the change
Properly fail the job when `SplitEnumeratorContext.runInCoordinatorThread()` throws an exception instead of failing the whole `JobManager`.


## Brief change log

Introduce `FailJobExceptionHandler` that calls `context.failJob()`. Call it when `SplitEnumeratorContext.runInCoordinatorThread()` throws an exception by changing `CoordinatorExecutorThreadFactory` error handler.
Introduce a new `ThreadPoolExecutor` test utility to gain access to the underlying thread and use it to wait for the handler to process the exception before asserting.

## Verifying this change

This change added tests and can be verified as follows:
`SourceCoordinatorContextTest#testExceptionInRunnableFailsTheJob()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes, fail the job instead of failing the JobManager when an exception is thrown
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no, bug fix
  - If yes, how is the feature documented? failure handling is documented in `SourceCoordinatorContext#runInCoordinatorThread`

R: @tillrohrmann 
Already squashed and split commits because some are related to `flink-test-utils` module.

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

tillrohrmann reviewed (changes requested):
Thanks a lot for this PR @echauchot. I think the PR looks already really good. The two comments I had were a suggestion for simplifying the test and a comment about the `FailJobExceptionHandler`. Once they are resolved, we can merge this PR.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/FailJobExceptionHandler.java:
```diff
@@ -0,0 +1,49 @@
+public final class FailJobExceptionHandler implements Thread.UncaughtExceptionHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FailJobExceptionHandler.class);
+    private final OperatorCoordinator.Context context;
```
of course, I totally agree it is a lot better, I should have thought about it ! Thanks for the suggestion

echauchot commented on flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java:
```diff
@@ -155,6 +156,41 @@ private void testAssignSplitToUnregisterdReader(boolean fromCoordinatorExecutor)
+        // wait that the thread handles the exception through the provided exception handler.
+        // Using ThreadPoolExecutor#awaitTermination is not enough to wait for exception handling.
+        // It just waits for the runnable
+        coordinatorExecutor.getSingleThread().join();
```
I agree it is simpler and higher level. I did not do that because I did not want to change the signature of the existing test components but fair enough, I'll do that. Thanks for the suggestion.

echauchot replied:
@tillrohrmann just a confirmation, you want `MockOperatorCoordinatorContext#failjob()` to become asynchronous? That would require passing the `ExecutorService` to `MockOperatorCoordinatorContext` and blocking on the Future termination in many tests. Maybe I misunderstood ?

echauchot replied:
Makes total sense, thanks Till !

echauchot replied:
That being said, master as moved `coordinatorExecutor` to be `ScheduledExecutorService` which catches the exceptions in `ScheduledFutureTask` so the exception handling works differently now. So I need to adapt the production code as well.

echauchot commented:
@tillrohrmann 
master as moved `coordinatorExecutor` to be a `ScheduledExecutorService`. Now that `coordinatorExecutor` is a `ScheduledExecutorService`, `UncaughtExceptionHandler` is useless as the JVM now uses a `ScheduledFutureTask` that catches exceptions in its `run()` method. We can only get the task Exception through `ScheduledFutureTask#get()`. So `runInCoordinatorThread()` now blocks until the task is done to check its Exceptions and fail the job if there is one. Before it just did a send and forget (did not care about the result of the Future) and could not react to exceptions, so it did not fail the jobManager either.

PTAL

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:
```diff
@@ -257,9 +260,19 @@ public void signalNoMoreSplits(int subtask) {
     @Override
     public void runInCoordinatorThread(Runnable runnable) {
-        coordinatorExecutor.execute(runnable);
+        final Future<?> future = coordinatorExecutor.submit(runnable);
+        try {
+            future.get(RUN_IN_COORDINATOR_THREAD_TIMEOUT, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            LOG.error(
+                    "Job failed: Thread '{}' produced an uncaught exception. Stopping the job...",
+                    coordinatorThreadName,
+                    e);
+            operatorCoordinatorContext.failJob(e);
+        }
```
I think we cannot do it like this because it will block the caller (main thread). I think we need to do something similar to https://stackoverflow.com/questions/12175985/how-to-make-scheduledthreadpool-report-errors where we wrap the runnable to catch the exception.

echauchot replied:
done, PTAL

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @echauchot. I think the blocking behaviour of `SourceCoordinatorContext.runInCoordinatorThread` is a bit problematic because it can block other components which is changing the contract of this method.

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @echauchot. The change itself looks good to me. However, there seem to be some test failures occurring now. I think we should iron them out before we can merge this PR.

echauchot commented:
> Thanks for updating this PR @echauchot. The change itself looks good to me. However, there seem to be some test failures occurring now. I think we should iron them out before we can merge this PR.

I took a look it seems to be unrelated memory issues on other tests.
`Feb 23 16:40:42 [ERROR] Picked up JAVA_TOOL_OPTIONS: -XX:+HeapDumpOnOutOfMemoryError
`
relaunching the build.

echauchot commented:
@flinkbot run azure

echauchot commented:
@flinkbot run azure

tillrohrmann commented:
I don't think that this is an unrelated issue:

```
"main" #1 prio=5 os_prio=0 tid=0x00007f2ef000b800 nid=0x7f58 waiting on condition [0x00007f2ef7cee000]
2022-02-24T12:35:37.4636644Z Feb 24 12:35:37    java.lang.Thread.State: WAITING (parking)
2022-02-24T12:35:37.4637310Z Feb 24 12:35:37 	at sun.misc.Unsafe.park(Native Method)
2022-02-24T12:35:37.4638464Z Feb 24 12:35:37 	- parking to wait for  <0x000000008965d770> (a java.util.concurrent.CompletableFuture$Signaller)
2022-02-24T12:35:37.4639357Z Feb 24 12:35:37 	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
2022-02-24T12:35:37.4640288Z Feb 24 12:35:37 	at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
2022-02-24T12:35:37.4641240Z Feb 24 12:35:37 	at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
2022-02-24T12:35:37.4642419Z Feb 24 12:35:37 	at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
2022-02-24T12:35:37.4643361Z Feb 24 12:35:37 	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
2022-02-24T12:35:37.4644495Z Feb 24 12:35:37 	at org.apache.flink.runtime.source.coordinator.SourceCoordinatorContextTest.testExceptionInRunnableFailsTheJob(SourceCoordinatorContextTest.java:188)
2022-02-24T12:35:37.4645536Z Feb 24 12:35:37 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2022-02-24T12:35:37.4646572Z Feb 24 12:35:37 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2022-02-24T12:35:37.4647550Z Feb 24 12:35:37 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2022-02-24T12:35:37.4648440Z Feb 24 12:35:37 	at java.lang.reflect.Method.invoke(Method.java:498)
2022-02-24T12:35:37.4649316Z Feb 24 12:35:37 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
2022-02-24T12:35:37.4650301Z Feb 24 12:35:37 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
2022-02-24T12:35:37.4651259Z Feb 24 12:35:37 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
2022-02-24T12:35:37.4652520Z Feb 24 12:35:37 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
2022-02-24T12:35:37.4653478Z Feb 24 12:35:37 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
2022-02-24T12:35:37.4654595Z Feb 24 12:35:37 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
2022-02-24T12:35:37.4655490Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2022-02-24T12:35:37.4656422Z Feb 24 12:35:37 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
2022-02-24T12:35:37.4657368Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
2022-02-24T12:35:37.4658301Z Feb 24 12:35:37 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
2022-02-24T12:35:37.4659295Z Feb 24 12:35:37 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
2022-02-24T12:35:37.4660194Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
2022-02-24T12:35:37.4661064Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
2022-02-24T12:35:37.4662260Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
2022-02-24T12:35:37.4663149Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
2022-02-24T12:35:37.4664035Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
2022-02-24T12:35:37.4664910Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
2022-02-24T12:35:37.4665731Z Feb 24 12:35:37 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
2022-02-24T12:35:37.4666510Z Feb 24 12:35:37 	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
2022-02-24T12:35:37.4667264Z Feb 24 12:35:37 	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
2022-02-24T12:35:37.4668079Z Feb 24 12:35:37 	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
2022-02-24T12:35:37.4668934Z Feb 24 12:35:37 	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
2022-02-24T12:35:37.4669760Z Feb 24 12:35:37 	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
2022-02-24T12:35:37.4670662Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
2022-02-24T12:35:37.4671636Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
2022-02-24T12:35:37.4672757Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
2022-02-24T12:35:37.4673419Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator$$Lambda$154/231977479.accept(Unknown Source)
2022-02-24T12:35:37.4674061Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
2022-02-24T12:35:37.4674913Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
2022-02-24T12:35:37.4675740Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
2022-02-24T12:35:37.4676333Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
2022-02-24T12:35:37.4676966Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
2022-02-24T12:35:37.4677646Z Feb 24 12:35:37 	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
2022-02-24T12:35:37.4678303Z Feb 24 12:35:37 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:188)
2022-02-24T12:35:37.4678974Z Feb 24 12:35:37 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:154)
2022-02-24T12:35:37.4679626Z Feb 24 12:35:37 	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:124)
2022-02-24T12:35:37.4680358Z Feb 24 12:35:37 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
2022-02-24T12:35:37.4680949Z Feb 24 12:35:37 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2022-02-24T12:35:37.4681486Z Feb 24 12:35:37 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
2022-02-24T12:35:37.4682272Z Feb 24 12:35:37 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
```

echauchot commented:
> ime.source.coordinator.SourceCoordinatorContextTest.testExceptionInRunnableFailsTheJob(SourceCoordinatorContextTest.java:188)

Oops, I missed this thread dump ! You're right it is definitely linked. 
Thanks for pointing out Till !

echauchot commented:
@tillrohrmann back on Flink, trying to reproduce the thread park

echauchot commented:
Funny, `SourceCoordinatorContextTest#testExceptionInRunnableFailsTheJob` run alone passes but when run with all the tests in the test class, it fails to compete the future. There must be a static somewhere in the test class, investigating...

dmvk commented:
Hi @echauchot, just FYI, since yesterday was a last day for @tillrohrmann, I'll take over the review from here.

echauchot commented:
Hey @dmvk thanks !

echauchot commented:
@tillrohrmann @dmvk. Fixed. For the record I had to use `ManuallyTriggeredScheduledExecutorService` to manually trigger runnables that ScheduledThreadPool keeps in a waiting state that caused the infinite wait. I had to modify `ManuallyTriggeredScheduledExecutorService` to allow passing the fthread factory so that the exception handler is set.

Already squashed and rebased. I think we are ready for merging.

PTAL

echauchot commented:
@flinkbot run azure

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:
```diff
@@ -257,9 +257,23 @@ public void signalNoMoreSplits(int subtask) {
+            try {
+                runnable.run();
+            } catch (final Throwable t) {
+                // when using a SheduledThreadPool, uncaught exception handler catches only
+                // exceptions thrown by the threadPool, so manually call it when the exception is
+                // thrown by the runnable
+                coordinatorThreadFactory.uncaughtException(Thread.currentThread(), t);
```
I don't think I understand this. The provided `Runnable` should be always executed by the thread from the pool. Any uncaught exception from the `Runnable` should be passed to the uncaught exception handler of the thread pool.

Why do we need this safeguard? Is there anything I'm missing?

echauchot replied:
See the history in the PR. The pb is that now that coordinatorExecutor is a ScheduledExecutorService, UncaughtExceptionHandler is useful only for exception raised by the threadPool as the JVM now uses a ScheduledFutureTask that catches exceptions in its run method. So the exception handler is no more for the runnable exceptions but for exceptions raised by the pool itself. It is also called in a manual try/catch in the runnable execution in `wrap()` (that Till recommended) to workaround the pb with ScheduledFutureTask explained above.

dmvk replied:
👍 Makes sense.

I'm wondering why are we using the ScheduledExecutor here at first place 🤔 

```
        final Executor errorHandlingCoordinatorExecutor =
                (runnable) ->
                        coordinatorExecutor.execute(
                                new ThrowableCatchingRunnable(
                                        this::handleUncaughtExceptionFromAsyncCall, runnable));

        this.notifier = new ExecutorNotifier(workerExecutor, errorHandlingCoordinatorExecutor);
```

We're effectively transforming it into the `Executor` anyway. Would simply switching from `Executors.newScheduledThreadPool` to `Executors.newSingleThreadExecutor` do the trick?

echauchot replied:
Yes I wondered also. But I'm just reacting to a change on master during the review (no luck). In the first version I indeed used `Executors.newSingleThreadExecutor` it worked well but it no longer complies to the signature of the context method.

becketqin replied:
The interface was changed due to [FLIP-182](https://cwiki.apache.org/confluence/display/FLINK/FLIP-182%3A+Support+watermark+alignment+of+FLIP-27+Sources), which schedules a watermark synchronization runnable in the coordinator executor. And the `scheduleAtFixedRate()` method resides in `ScheduledExecutorService`.

dmvk replied:
OK, I see that now, I've missed the `org.apache.flink.runtime.source.coordinator.SourceCoordinatorContext#getCoordinatorExecutor` 🤦 

Thank you both 👍

dmvk commented on flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java:
```diff
@@ -155,6 +156,40 @@ private void testAssignSplitToUnregisterdReader(boolean fromCoordinatorExecutor)
+        ManuallyTriggeredScheduledExecutorService manualWorkerExecutor =
+                new ManuallyTriggeredScheduledExecutorService();
+        // need the factory to have the exception handler set
+        final ManuallyTriggeredScheduledExecutorService coordinatorExecutorWithExceptionHandler =
+                new ManuallyTriggeredScheduledExecutorService(1, coordinatorThreadFactory);
```
This implementation executes `Runnables` directly in the thread that has executed the `trigger...` method, so the exception handler is basically ignored.

I guess this might have been the reason for introducing the `wrap` method on the context, otherwise this test shouldn't pass.

Also the changes to `ManuallyTriggeredScheduledExecutorService` don't seem to have any effect, so we should be able to get rid of them.

Can we do something as simple as ... ?

```java
        final ScheduledExecutorService executorService =
                Executors.newSingleThreadScheduledExecutor(
                        new SourceCoordinatorProvider.CoordinatorExecutorThreadFactory("test", operatorCoordinatorContext));
```

Another option would be adding support for the exception handler to `ManuallyTriggeredScheduledExecutorService` (basically wrapping `next.run();` in the `trigger` method with `try/catch`.

echauchot replied:
see my previous comment above it is only for exceptions in the threadpool itself.

echauchot replied:
another comment, `wrap()` was introduced in the PR before using `ManuallyTriggeredScheduledExecutorService` when `Executors.newSingleThreadScheduledExecutor` was in use. It was introduced because `ScheduledFutureTask` catches its own exceptions.

dmvk commented on flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/ManuallyTriggeredScheduledExecutorService.java:
```diff
@@ -54,6 +57,14 @@
+    }
+
+    public ManuallyTriggeredScheduledExecutorService(int nbThreads, ThreadFactory factory) {
+        super(nbThreads, factory);
```
This change IMO doesn't have any effect. Can you please elaborate?

echauchot replied:
except set the exception handler in ManuallyTriggeredScheduledExecutorService

dmvk replied:
since we override the `execute(...)` method (which is the only one used by the coordinator), how exactly do we leverage that?

echauchot replied:
`CoordinatorExecutorThreadFactory#newThread()` sets the exception handler of the created thread.   The JVM calls `Thread#getUncaughtExceptionHandler()` when the threads raises an exception.

dmvk replied:
That part I understand. What I miss is where do we actually create the thread (call the `#newThread` method), since we seem to be executing the `Runnable` passed to the `execute(...)` directly?

echauchot replied:
you're right, the context as a reference to the factory and calls the exception handler directly. No more need for that change, reverted.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/FailJobExceptionHandler.java:
```diff
@@ -0,0 +1,49 @@
+ * <p>This guarantees that critical exceptions are not accidentally lost and leave the system
+ * running in an inconsistent state.
+ */
+public final class FailJobExceptionHandler implements Thread.UncaughtExceptionHandler {
```
Name is a leftover of the time it actually called the `context.failjob()` now that we changed it to `Cosumer<Throwable>` I agree it is misleading and BTW it is a class not a lambda because the is already exception handlers in the code base so I wanted to be consistent.

echauchot replied:
If you prefer, as `FailJobExceptionHandler` is only a `Consumer<Throwable>` in the end, I can make this class a lambda `(Thread t, Throwable e) -> {context.failjob()}`. WDYT ?

echauchot replied:
As it was what you suggested in the first place, I did the change above

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/FailJobExceptionHandler.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public void uncaughtException(Thread t, Throwable e) {
+        LOG.error(
+                "Job failed: Thread '{}' produced an uncaught exception. Stopping the job...",
```
:+1

echauchot replied:
done

dmvk commented on flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java:
```diff
@@ -257,9 +257,23 @@ public void signalNoMoreSplits(int subtask) {
     @Override
     public void runInCoordinatorThread(Runnable runnable) {
-        coordinatorExecutor.execute(runnable);
+        coordinatorExecutor.execute(wrap(runnable));
```
This seems similar to what `errorHandlingCoordinatorExecutor` already does. Could we reuse it here?

echauchot replied:
I don't see this class, where is it ?

dmvk replied:
https://github.com/apache/flink/blob/5e4efadcae3f89d0661470b754f920a3f0cae337/flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java#L137

echauchot replied:
thanks. I guess you mean using `ThrowableCatchingRunnable` with the coordinatorExecutor. I totally agree.
Done.

dmvk reviewed (commented):
Thanks @echauchot for the discussion, I think I have most of the context I've been missing. 👍 

Few questions:
- Do we really need changes to the ManualScheduledExecutor.. ? It doesn't seem to create any threads, because all runnables are executed directly by calling `trigger...`.
- Do we also need to handle failures in `SourceCoordinator#announceCombinedWatermark`?
- We already have the `errorHandlingCoordinatorExecutor`. Could that be reused instead of creating another wrapper?

echauchot commented:
@flinkbot run azure

echauchot commented:
> Thanks @echauchot for the discussion, I think I have most of the context I've been missing. +1
> 
> Few questions:
> 
> * Do we really need changes to the ManualScheduledExecutor.. ? It doesn't seem to create any threads, because all runnables are executed directly by calling `trigger...`.
=> agree
> * Do we also need to handle failures in `SourceCoordinator#announceCombinedWatermark`?
=> I guess it is outside of the scope of this PR. But if needed, I could do a second PR.
> * We already have the `errorHandlingCoordinatorExecutor`. Could that be reused instead of creating another wrapper?
=> done

echauchot commented:
@dmvk thanks for your review ! I addressed all your comments and squashed the review commits. Is the PR ready to be merged ?

echauchot commented:
@dmvk failure on build is a flaky test that should be solved by [one of my ongoing PR](https://github.com/apache/flink/pull/18973)

echauchot commented:
@flinkbot run azure

dmvk reviewed (approved):
LGTM 👍 Thanks for the contribution @echauchot.

dmvk commented:
@flinkbot run azure

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java:
```diff
@@ -155,6 +156,40 @@ private void testAssignSplitToUnregisterdReader(boolean fromCoordinatorExecutor)
+                });
+
+        manualWorkerExecutor.triggerAll();
+        // testingContext.close(); already shutdowns coordinatorExecutor and blocks until tasks are
+        // finished
```
```suggestion
        // already shutdowns coordinatorExecutor and blocks until tasks are finished
```

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContextTest.java:
```diff
@@ -155,6 +156,40 @@ private void testAssignSplitToUnregisterdReader(boolean fromCoordinatorExecutor)
+        // finished
+        testingContext.close();
+        coordinatorExecutorWithExceptionHandler.triggerAll();
+        // blocks until the job is failed :
+        // wait that the uncaught exception handler calls operatorCoordinatorContext#failJob which
+        // completes the future
```
```suggestion
        // blocks until the job is failed: wait that the uncaught exception handler calls
        // operatorCoordinatorContext#failJob which completes the future
```

XComp reviewed (approved):
I'm going to merge the PR considering that @dmvk took over the review from @tillrohrmann. Thanks for you contribution, @echauchot. I'd like to ask you to squash the commits into a single commit. Feel free to put parts of the individual commit messages into the resulting commit message if it makes sense.

I did a bit of nit-picking with some of the comments (just cosmetic stuff). It would be nice to have these fixed as well when squashing the changes into a single commit. Please check my comments below.

echauchot commented:
@XComp thanks for your review. I addressed your comments and squashed as requested. PTAL

echauchot commented:
@flinkbot run azure

XComp commented:
Thanks for your contribution @echauchot . I created backport PRs for your this one to cover the issue also in older versions. I will merge this PR next.

echauchot commented:
@XComp thanks for merging !
